src/compiler/checker.ts

```diff
@@ -8111,20 +8144,22 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             // we write it out like that, rather than as
             // export const x: <T>(x: T) => T
             // export const y: <T_1>(x: T_1) => T_1
-            if (initial.typeParameterNames) {
-                initial.typeParameterNames = new Map(initial.typeParameterNames);
-            }
-            if (initial.typeParameterNamesByText) {
-                initial.typeParameterNamesByText = new Set(initial.typeParameterNamesByText);
-            }
-            if (initial.typeParameterSymbolList) {
-                initial.typeParameterSymbolList = new Set(initial.typeParameterSymbolList);
-            }
-            if (initial.typeParameterNamesByTextNextNameCount) {
-                initial.typeParameterNamesByTextNextNameCount = new Map(initial.typeParameterNamesByTextNextNameCount);
-            }
-            initial.tracker = new SymbolTrackerImpl(initial, initial.tracker.inner, initial.tracker.moduleResolverHost);
-            return initial;
+            const oldMustCreateTypeParameterSymbolList = context.mustCreateTypeParameterSymbolList;
+            const oldMustCreateTypeParametersNamesLookups = context.mustCreateTypeParametersNamesLookups;
+            context.mustCreateTypeParameterSymbolList = true;
+            context.mustCreateTypeParametersNamesLookups = true;

```

Comment:
Is all of this `mustCreate*` stuff an optimization? The logic behind it isn't super clear to me.

Reply:
Yes. It's an optimization. It's basically copy on write. When creating a new context for a signature it's not always necessary to also duplicate the lookups. This defers the copy until there is actually a reason copy (until there is mutation of the map). In both of these cases (mustCreateTypeParameterSymbolList  and mustCreateTypeParametersNamesLookups) the mutation of the map only occurs in one place so it's easy enough to enforce this. 

The optimizations in the last comit avoids duplicating the context, the symbol tracker as well as the type parameters lookup maps (if the signature is not generic).